File: Serenity.Script.UI/Reporting/ReportPage.cs
namespace Serenity.Reporting
{
    using jQueryApi;

    public class ReportPage : Widget
    {
        public ReportPage(jQueryObject div)
            : base(div)
        {
            J(".report-link").Click(ReportLinkClick);

            J("div.line").Click(CategoryClick);

            var self = this;

            new QuickSearchInput(J("#QuickSearchInput"), new QuickSearchInputOptions
            {
                OnSearch = (field, text) =>
                {
                    self.UpdateMatchFlags(text);
                }
            });
        }

        private void UpdateMatchFlags(string text)
        {
            var liList = J("#ReportList").Find("li").RemoveClass("non-match");

            text = text.TrimToNull();
            if (text == null)
            {
                liList.Children("ul").Hide();
                liList.Show().RemoveClass("expanded");
                return;
            }

            var parts = text.Split(new char[] { ',', ' ' }, System.StringSplitOptions.RemoveEmptyEntries);
            for (var i = 0; i < parts.Length; i++)
                parts[i] = Q.Externals.StripDiacritics(parts[i]).ToUpperCase().TrimToNull();

            var reportItems = liList.Filter(".report-item");
            
            reportItems.Each((i, e) =>
            {
                var x = J(e);
                var title = Q.Externals.StripDiacritics((x.GetText() ?? "").ToUpperCase());

                foreach (var p in parts)
                    if (p != null && !title.Contains(p))
                    {
                        x.AddClass("non-match");
                        break;
                    }
            });
            
            var matchingItems = reportItems.Not(".non-match");
            var visibles = matchingItems.Parents("li").Add(matchingItems);
            var nonVisibles = liList.Not(visibles);
            nonVisibles.Hide().AddClass("non-match");
            visibles.Show();

            if (visibles.Length <= 100)
            {
                liList.Children("ul").Show();
                liList.AddClass("expanded");
            }
        }

        private void CategoryClick(jQueryEvent e)
        {
            var li = J(e.Target).Closest("li");
            if (li.HasClass("expanded"))
            {
                li.Find("ul").Hide(EffectDuration.Fast);
                li.RemoveClass("expanded");
                li.Find("li").RemoveClass("expanded");
            }
            else
            {
                li.AddClass("expanded");
                li.Children("ul").Show(EffectDuration.Fast);
                if (li.Children("ul").Children("li").Length == 1 &&
                    !li.Children("ul").Children("li").HasClass("expanded"))
                    li.Children("ul").Children("li").Children(".line").Click();
            }
        }

        private void ReportLinkClick(jQueryEvent e)
        {
            e.PreventDefault();

            var dialog = new ReportDialog(new ReportDialogOptions
            {
                ReportKey = J(e.Target).GetDataValue("key").As<string>()
            });
        }
    }
}

File: common-features/src/extensions/Modules/Reporting/ReportPage.ts
import { Fluent, QuickSearchInput, Widget, WidgetProps, stripDiacritics } from "@serenity-is/corelib";
import { bindThis } from "@serenity-is/domwise";
import { nsExtensions } from "../ServerTypes/Namespaces";
import { ReportDialog } from "./ReportDialog";

export class ReportPage<P = {}> extends Widget<P> {
    static override[Symbol.typeInfo] = this.registerClass(nsExtensions);

    constructor(props: WidgetProps<P>) {
        super(props);

        Fluent.on(this.domNode, "click", ".report-link", bindThis(this).reportLinkClick);

        new QuickSearchInput({
            element: this.domNode.querySelector('.s-QuickSearchBar input') as HTMLElement,
            onSearch: (field, text, done) => {
                this.updateMatchFlags(text);
                done(true);
            }
        });
    }

    protected updateMatchFlags(text: string) {
        const liList = this.domNode.querySelectorAll('.report-list li');
        liList.forEach(x => x.classList.remove('non-match'));
        text = text?.trim();
        if (!text)
            return;

        text = stripDiacritics(text).toUpperCase();

        const reportItems = Array.from(liList).filter(x => x.classList.contains('report-item'));
        reportItems.forEach(function (el) {
            const title = stripDiacritics((el.textContent ?? '').toUpperCase());
            if (title.indexOf(text) < 0) {
                el.classList.add('non-match');
            }
        });

        function parents(el: HTMLElement, selector: string) {
            const parents = [];
            while ((el = el.parentNode as HTMLElement) && (el as any) !== document) {
                if (!selector || el.matches(selector))
                    parents.push(el);
            }
            return parents;
        }

        const matchingItems = reportItems.filter(x => !x.classList.contains('non-match'));
        let visibles = [...matchingItems];
        matchingItems.forEach(x => visibles.push(...parents(x as HTMLElement, 'li')));
        visibles = visibles.filter((x, i) => visibles.indexOf(x) === i);
        visibles.forEach(v => {
            v.querySelectorAll(':scope [data-bs-toggle]:not([aria-expanded=true])').forEach(c => {
                c.setAttribute('aria-expanded', "true");
                c.classList.remove('collapsed');
            });
            if (v.parentElement && v.parentElement.classList.contains("collapse") &&
                !v.parentElement.classList.contains("show")) {
                v.parentElement.classList.add("show");
            }
        });

        const nonVisibles = Array.from(liList).filter(x => visibles.indexOf(x) < 0);
        nonVisibles.forEach(x => x.classList.add('non-match'));
    }

    protected reportLinkClick(e: Event) {
        e.preventDefault();
        new ReportDialog({
            reportKey: (e.target as HTMLElement).getAttribute('data-key')
        }).dialogOpen();
    }
}
